experimental/normalizer/src/lib.rs

```diff
@@ -0,0 +1,448 @@
+/// be set and not on a temporary copy.
+///
+/// Note: As a micro-optimization, this struct is
+/// distinct from the struct of the same name in
+/// the collator. They have the opposite default
+/// bit pattern for the high 8 bits.
+#[derive(Debug)]
+struct CharacterAndClass(u32);
+
+impl CharacterAndClass {
+    pub fn new(c: char) -> Self {
+        // The combining class bits default to zero.
+        CharacterAndClass(u32::from(c))
+    }
+    pub fn character(&self) -> char {
+        unsafe { char::from_u32_unchecked(self.0 & 0xFFFFFF) }
+    }
+    pub fn ccc(&self) -> CanonicalCombiningClass {
+        CanonicalCombiningClass((self.0 >> 24) as u8)
+    }
+    // XXX need better naming here.
+    pub fn set_ccc(&mut self, ccc: &CodePointTrie<CanonicalCombiningClass>) {
+        debug_assert_eq!(self.0 >> 24, 0, "This method has already been called!");
+        self.0 |= (ccc.get(self.0).0 as u32) << 24;
+    }
+}
+
+// This function exists as a borrow check helper.
+#[inline(always)]
+fn sort_slice_by_ccc<'data>(
+    slice: &mut [CharacterAndClass],
+    ccc: &CodePointTrie<'data, CanonicalCombiningClass>,
+) {
+    // We don't look up the canonical combining class for starters
+    // of for single combining characters between starters. When
+    // there's more than one combining character between starters,
+    // we look up the canonical combining class for each character
+    // exactly once.
+    if slice.len() < 2 {
+        return;
+    }
+    slice.iter_mut().for_each(|cc| cc.set_ccc(ccc));
+    slice.sort_by_key(|cc| cc.ccc());
+}
+
+pub struct Decomposition<'data, I>
+where
+    I: Iterator<Item = char>,
+{
+    delegate: I,
+    buffer: SmallVec<[CharacterAndClass; 10]>, // TODO Figure out good length
+    pending_unnormalized_starter: Option<char>, // None at end of stream
+    decompositions: &'data CanonicalDecompositionDataV1<'data>,
+    ccc: &'data CodePointTrie<'data, CanonicalCombiningClass>,
+}
+
+impl<'data, I> Decomposition<'data, I>
+where
+    I: Iterator<Item = char>,
+{
+    pub fn new(
+        delegate: I,
+        decompositions: &'data CanonicalDecompositionDataV1,
+        ccc: &'data CodePointTrie<'data, CanonicalCombiningClass>,
+    ) -> Self {
+        let mut ret = Decomposition::<I> {
+            delegate: delegate,
+            buffer: SmallVec::new(), // Normalized
+            // Initialize with a placeholder starter in case
+            // the real stream starts with a non-starter.
+            pending_unnormalized_starter: Some('\u{FFFF}'),
+            decompositions: decompositions,
+            ccc: ccc,
+        };
+        let _ = ret.next(); // Remove the U+FFFF placeholder
+        ret
+    }
+}
+
+impl<'data, I> Iterator for Decomposition<'data, I>
+where
+    I: Iterator<Item = char>,
+{
+    type Item = char;
+
+    fn next(&mut self) -> Option<char> {
+        if !self.buffer.is_empty() {
+            return Some(self.buffer.remove(0).character());
```

Comment:
The buffer is filled and then drained completely before being filled again, so it doesn't need to be circular. I added an extra index of state to decouple the logical read advance from the start of the buffer so that the remaining items don't need to be moved.

And, yes, it needs to be able to grow to an arbitrary size, because Unicode doesn't limit the number of combining marks.

The ICU4C normalizer is structurally very different.